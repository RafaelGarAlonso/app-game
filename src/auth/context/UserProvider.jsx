import { UserContext } from './UserContext';
import { INIT_USER, INIT_USERS_TABLE } from '../../constants/users'

export const UserProvider = ({ children }) => {

  const getUser = () => {
    const localStorageData = localStorage.getItem('currentUser');
    return localStorageData ? JSON.parse(localStorageData) : INIT_USER;
  };

  const setNewUserSession = ({ userName }) => {
    const listOfUsers = getListOfUsers();

    for (const [index, user] of listOfUsers.entries()) {
      if (user.userName === userName) {
        const { userName, points } = listOfUsers[index];
        localStorage.setItem('currentUser', JSON.stringify({ userName, points }));
        return false;
      }
    };

    localStorage.setItem('currentUser', JSON.stringify({ userName, points: 0 }));
    return true;
  };

  const getAppLoaded = () => {
    return sessionStorage.getItem('appLoaded');
  };

  const setAppLoaded = (loaded) => {
    sessionStorage.setItem('appLoaded', loaded);
  };

  const userEarnPoints = () => {
    const user = getUser();
    localStorage.setItem('currentUser', JSON.stringify({ userName: user.userName, points: user.points + 1 }));
  };

  const getListOfUsers = () => {
    const listOfUsers = localStorage.getItem('listOfUsers');
    return listOfUsers ? JSON.parse(listOfUsers) : null;
  };

  const setNewListOfUsers = (listOfUsers = INIT_USERS_TABLE) => {
    localStorage.setItem('listOfUsers', JSON.stringify(listOfUsers));
  };

  const setPointsToListUsers = (currentUser) => {
    const listOfUsers = getListOfUsers();

    for (let [index, user] of listOfUsers.entries()) {
      if (user.userName === currentUser.userName) {
        listOfUsers[index].points = currentUser.points;
        break;
      }
    };

    setNewListOfUsers(listOfUsers);
  };

  return (
    <UserContext.Provider value={{
      setNewUserSession,
      getUser,
      setAppLoaded,
      getAppLoaded,
      userEarnPoints,
      getListOfUsers,
      setNewListOfUsers,
      setPointsToListUsers
    }}>
      {children}
    </UserContext.Provider>
  )
}