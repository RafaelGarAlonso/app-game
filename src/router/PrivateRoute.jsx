import { useContext } from 'react';
import { Navigate } from 'react-router-dom';
import { UserContext } from '../auth/context/UserContext';

export const PrivateRoute = ({ children }) => {
    const { getUser } = useContext(UserContext);
    const { userName } = getUser();

    if (!userName || userName === "") return <Navigate to="/home" />

    return children;
}