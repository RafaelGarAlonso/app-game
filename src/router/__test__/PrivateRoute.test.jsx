import { render } from '@testing-library/react';
import { UserProvider } from '../../auth/context/UserProvider';
import { PrivateRoute } from '../PrivateRoute';
import { MemoryRouter } from 'react-router-dom';

describe('Test <PrivateRoute />', () => {

    const renderTemplate = () => {
        return render(
            <UserProvider>
                <MemoryRouter>
                    <PrivateRoute />
                </MemoryRouter>
            </UserProvider>
        );
    }

    test('Check snapshot component', () => {
        const { container } = renderTemplate();
        expect(container).toMatchSnapshot();
    });
});