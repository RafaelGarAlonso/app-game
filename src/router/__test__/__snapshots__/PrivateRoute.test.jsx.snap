// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <PrivateRoute /> Check snapshot component 1`] = `<div />`;
